### ISSUE TITLE
Add feature to follow latest release

### ISSUE THREAD

**Issue by mjlbach** [wbthomason/packer.nvim#797]
Many users are complaining about package instability. One solution for plugins is to follow semantic versioning. It would be nice to have an option, in packer, to follow the latest tagged release of a plugin.

The alternative, would be to establish a convention for plugins in which releases were pushed to a release branch.

**Comment by shadmansaleh** [comment]
duplicate #529 😁

**Comment by mjlbach** [comment]
While related, this isn't a duplicate. Having wildcard support for tags is a different design choice.

I'm suggesting an explicit mechanism around releases, with possible conventions around warning users on non-minor version updates.

**Comment by shadmansaleh** [comment]
> I'm suggesting an explicit mechanism around releases
with possible conventions around warning users on non-minor version updates.

Okey . Warning users is a good idea

**Comment by danymat** [comment]
This is a very good idea and I think it can provide a great way for users to update to latest stable release in Neorg. If this works, I will try to port it to Neogen too

**Comment by danymat** [comment]
@wbthomason Sorry for pinging you, but as a plugin contributor I'm fairly sure this could be a very welcomed addition for users. Any feedback on this ?

**Comment by wbthomason** [comment]
To make sure I understand: this would involve finding the latest commit tagged `release` or similar? I think that would be good to add, yes.

**Comment by wbthomason** [comment]
If my interpretation is correct, then - does anyone know a convenient way to do this with `git`? I can figure it out, but don't know a command off of the top of my head.

**Comment by danymat** [comment]
I don't think that's this. I think @mjlbach meant about following releases from the `Release` section of a plugin that is tagged

**Comment by wbthomason** [comment]
@danymat Sorry, I'm not sure what the "Release section of a plugin" is?

**Comment by wbthomason** [comment]
If you mean Github's "Releases" page, then I'm trying to figure out how to list those with `git`.

**Comment by danymat** [comment]
For example, in neorg we used a tag called `latest`, that basically was updated with latest tag that was created.
I.e:
```
latest -- points to 1.8 tag commit
1.8
1.7
```

new version comes out:

```
latest -- points to 1.9 tag commit
1.9
1.8
1.7
```

The problem here is double: as @mrossinek suggested, changing a tag commit after a push is not good practice for users, because that make them unaware if the tag is a tag that changes or not.

So we removed the `latest` tag, and instruct the users to use whichever versioned tag they want (for example: `1.8`).

Some idea of a feature could be to ask plugins to use semver tags (`1.7`, `1.8`), and have packer have a `follow_semver = "major"` that could update to the latest major tag, for example (see: https://semver.org for what a major version is)

**Comment by mjlbach** [comment]
I don't think you can since it's an artifact of github. Tags could be in some arbitrary order. You would need to query the github endpoint.

```bash
get_latest_release() {
  curl --silent "https://api.github.com/repos/$1/releases/latest" | # Get latest release from GitHub api
    grep '"tag_name":' |                                            # Get tag line
    sed -E 's/.*"([^"]+)".*/\1/'                                    # Pluck JSON value
}
```

**Comment by wbthomason** [comment]
Right, so this is sort of what I'm asking: do we want to introduce a convention for Github plugin authors of having a consistent "special" tag (e.g. `release`) that is expected to move, or do we want to introduce the Github-specific feature of querying the releases endpoint?

**Comment by wbthomason** [comment]
re: semver: I'm not against adding semver support to `packer`, but I'd prefer to add something like the `requirements.lua` format we've previously discussed, which would let us manage plugin and dependency versions together.

**Comment by danymat** [comment]
> "special" tag (e.g. release) that is expected to move

As i mentionned before, I think this ability could cause some problems for users, and is. generally considered bad practice, thus I don't think we should standardize this feature. I'm more found of a "semver" feature, or a requirements table indeed. 

One question though. Is the requirements table supposed to follow a commit, or a git tag?

**Comment by mjlbach** [comment]
One option (which would not require any modification to packer at all) would be, as I think @wbthomason is implying, to establish a convention (and provide a GHA that handles this automatically) of pushing the latest stable tagged GH release to a `release` branch.

**Comment by wbthomason** [comment]
^ That would be fine, but I was suggesting (as option 1) a tag that is _expected_ to move (so that users should not be caught by surprise).

I'm a bit skeptical of getting plugin authors broadly onboard with a `release` branch convention, but maybe that's easier than getting them onboard with adding a `requirements.lua` file or similar.

**Comment by danymat** [comment]
> ^ That would be fine, but I was suggesting (as option 1) a tag that is expected to move (so that users should not be caught by surprise).

Please read this, maybe this can help in the thought process: https://git-scm.com/docs/git-tag#_on_re_tagging

**Comment by mjlbach** [comment]
I think a tag vs. branch has essentially equivalent difficulty for the plugin authors (a GH release trigger push to either is easy to do, and we could provide the action in the neovim core organization) but the semantics of pushing to a release branch are cleaner (and more in line with convention) than overwriting the tag. The alternative would be querying the GH api which would increase the complexity of packer's core code (TBF not by much), and be a bit slower.

**Comment by wbthomason** [comment]
Sure. I see the options as:
1. `release` branch: no `packer` work, moderate plugin author work
2. Query GH: Github specific, slightly slower (roughly "1" network roundtrip for opted-in plugins) updates, moderate `packer` work, moderate plugin author work (to actually tag releases)
3. `requirements.lua`: solves other problems, moderate `packer` work, moderate plugin author work.

Given these, I favor (1) or (3). (3) is the most work of all for me, but the greatest benefit.

**Comment by mjlbach** [comment]
1 and 3 seem not mutually exclusive to me? I agree establishing a requirements.lua (or some other standardized conventions) has benefits that extend beyond this.

**Comment by danymat** [comment]
What about the semver tag convention ? Don't you think it's an option to consider too ?

**Comment by wbthomason** [comment]
> 1 and 3 seem not mutually exclusive to me? I agree establishing a requirements.lua (or some other standardized conventions) has benefits that extend beyond this.

Yeah, sorry - they're not exclusive, but I meant in terms of choosing what to do right now.

**Comment by wbthomason** [comment]
> What about the semver tag convention ? Don't you think it's an option to consider too ?

It's an option, but it seems to add the relative most work for both `packer` and plugin authors, no? I was considering (3) to subsume semver tags.

**Comment by mjlbach** [comment]
@wbthomason how would the requirements.lua specify the latest release? As a field specify the git sha, branch, or tag?

TBH my only concern right now is stopping the deluge of "neovim plugins are unstable because I used the latest git commit" reddit posts 😆

**Comment by wbthomason** [comment]
> TBH my only concern right now is stopping the deluge of "neovim plugins are unstable because I used the latest git commit" reddit posts

Absolutely. All of the solutions here seem to require convincing plugin authors to change behavior - which do you think will be the easiest? From the `packer` side of things, I'm happy to do any of them.

> @wbthomason how would the requirements.lua specify the latest release? As a field specify the git sha, branch, or tag?

Fair, I'm being stupid - we'd still need one of the other solutions to avoid redundant fetches (if e.g. the version was just in the `requirements.lua` file).

**Comment by mjlbach** [comment]
I think I am personally leaning towards `release` branch but I would like to hear from other notable plugin authors. @kristijanhusak do you all have any thoughts? (and sorry for looping you all in, but you should take it as an indication of how much I value your opinions 😆)

**Comment by lewis6991** [comment]
Quick reply.

> If we did have a convention for the package metadata we could include version bounds on the release and warn users if they were using a too-old neovim version.

This is the thing I care about the most.

I would also prefer the requirements.lua approach if we do anything, but may I suggest we just use rockspec files even if we're not using luarocks (...yet) as that format is already pretty standardized and should be able to capture any metadata we need.

**Comment by mjlbach** [comment]
Right, but even with the requirements.lua (I think it should be either plugins.lua since it goes beyond requirements, or if we go the rockspec direction following that naming convention) we still need a way to pull the latest release (either via tag patterns, embedding the commit in the plugins.lua, or via a release branch unless we start pushing all plugins to luarocks.

**Comment by danymat** [comment]
> we still need a way to pull the latest release (either via tag patterns, embedding the commit in the plugins.lua, or via a release branch unless we start pushing all plugins to luarocks).

Indeed, that's why the tag versioning (or the release branch) seems to be a good idea

**Comment by lewis6991** [comment]
You don't necessarily need to pull it, packer just needs to query GitHub if such a file exists and pull just that file if there is one? Should just be a couple of web requests?

**Comment by danymat** [comment]
> You don't necessarily need to pull it, packer just needs to query GitHub if such a file exists and pull just that file if there is one? Should just be a couple of web requests?

I don't think this is a possible "universal" convention, as some plugins could be on GitLab or other versioning platforms. But for a temporary solution it will surely work indeed

**Comment by mjlbach** [comment]
I didn't mean pull the `plugin.lua` file, I mean we need to specify what the latest release is somehow (even if information that points to the latest release was embedded in that `plugin.lua` file).

* Packer fetches the `plugin.lua`, parses it for the `latest_release` field, and then pulls the latest release because the commit and/or tag is stored in `plugin.lua`, I think this is a bit weird because plugin.lua is in itself versioned (so old plugin.lua will point to old latest-releases)
* There is a convention by which the latest release is stored, such as a `release branch` that packer can use explicitly.

The important question for *this* issue is what the source of truth is for the `latest` release.

**Comment by lewis6991** [comment]
> - Packer fetches the plugin.lua, parses it for the latest_release field

I don't think it needs to be anything more than that.

A `release-*` tag/branch will work too. I don't see why we can't support both (as been mentioned).

> I just want to avoid what happened the last time I proposed standardizing plugin specifications, is it ended up being an "all or nothing" situation in which nothing happened because we couldn't agree on an entire set of conventions, that's why I want to divorce the conversation about conventions for querying GH for the latest release from the question of package metadata standardization.

As you've said in the past, someone just needs to champion this. Once packer has something and some big plugins follow, everyone else will follow suite.

For Gitsigns I've already added: a rockspec,  a `release` tag/branch and a GitHub release. If there is some other file a plugin manager will parse, I'll add that too. Problem is, plugin managers haven't decided on anything.

**Comment by mjlbach** [comment]
>  Packer fetches the plugin.lua, parses it for the latest_release field

> I don't think it needs to be anything more than that.

I just think that is a bit of a strange convention, as the `latest_release` field is going to depend on the version of the `plugin.lua` file you are fetching. 

> A release-* tag/branch will work too. I don't see why we can't support both (as been mentioned).

The difference here being that I am proposing an un-versioned/rolling `release` branch that will always carry the latest tagged release. Plugins can still tag releases or create release branches (release-0.6, etc.) for pinning reasons, but this would make it easy to directly fetch the latest release (which would include the plugins.lua, which could be used to fetch dependencies). 

Where I see this becoming problematic is if `plugin.lua` specified version constraints that then needed to be resolved.

**Comment by lewis6991** [comment]
> I just want to avoid what happened the last time I proposed standardizing plugin specifications

I feel like I've caused this again...

Let's just go with something simple.

I propose these rolling/moving tags.

- `release`: latest release
- `release-0.5compat`: latest release that supports Neovim 0.5
- `release-<VERSION>compat`: latest release that supports Neovim `<VERSION>`

EDIT: alternative `release/compat/<VERSION>`

Packer can then choose which tag to pick up depending on the users Neovim version; should be pretty simple.

This achieves pretty much everything I would personally want for now.

**Comment by wbthomason** [comment]
If we go with a branch/tag, `packer` can add a `follow_latest_releases` config var which will auto-add the `branch = 'release'` constraint (or similar). This might help the convention gain adoption/can still be overridden if users want a different `compat` branch, etc.

EDIT: it would also be easy to have `packer` fetch the latest compatible branch if plugin authors adhere to a consistent format for those.

**Comment by mjlbach** [comment]
+1 to @lewis6991's proposal, I'm happy to discuss rockspec but I think this will (for now) address 90% of the issue. I would go with release-0.6 (supports neovim 0.6) and release (rolling)

**Comment by wbthomason** [comment]
This sounds good to me. Do we think that the proposed `packer` feature to auto-follow the release tag or branch (@lewis6991, you prefer tags? @danymat, are you OK with that?) would be worth adding to help the convention gain adoption?

**Comment by danymat** [comment]
@wbthomason im okay with autoflowing a branch
However in practice I'm not even sure if it's possible to pull annex change to a gut tag

**Comment by kristijanhusak** [comment]
I agree with most of the things above. We can take any route, but it needs to be consistent. My opinion would be to do what most of other package managers in other languages do: Follow semver and use tags.  I'm not sure how hard would be for packer to parse the wildcards, but I'm hoping not too much. For starters it might be slow, but if there is a long term plan to introduce some kind of version locking, it can be optimized. This is how I imagine it:

Plugin has these versions: v0.1, v0.2, v0.3, v1.0, v1.1, v1.2

1. Bleeding edge, which is basically the current functionality (default branch), would be marked as `version = 'latest'`. This could maybe be globally configurable for people who are adventurous and doesn't have the problem with potential package breakage on every update. It would remove the need to set this "latest" version on every plugin configuration. Maybe even have this option enabled by default, and have this versioning opt-in for now, so people can test it.

2. Using latest tag would be `version = '*'`.  This could become the default at some point. Since most of the plugins do not have any tags (shame on me), we could just default to the "latest" if there isn't any. This would avoid breaking current functionality to some extent.

3. Using specific minor version would be `version = '1.*'`.  Not much to explain here. Just a standard wildcard matching.

Ideally, Neovim community or Packer could provide a (simple for now) CLI that would allow creating these releases in an expected way. For example, if I have a release to put out, I could run a CLI command `packer release major/minor/patch` that would know how to increase the tag and publish it. That way we keep it consistent for all. Later, if we ever manage to switch to Luarocks or something else, we can extend the CLI to do more things. No need to introduce another tool at that moment. I know this is very ambitious at this point, but to me it sounds like a good way to go.

Biggest concern for me would be how to enforce this, especially for the plugins that are not Neovim specific. Vim community seems perfectly fine with the current functionality, at least from what I follow. All of this needs to be backward compatible to avoid frustration among the "average" Neovim users.

**Comment by danymat** [comment]
After @kristijanhusak I went to try and add support to the wildcard thing. You can check out #809 for implementation

**Comment by danymat** [comment]
Hello plugin devs, just for you to know: https://github.com/wbthomason/packer.nvim/pull/809 is merged, so that means if you use tags for releasing your plugins, you can use the willdcard to allow for updates, for example:

```lua
tag = "*" -- Will update last created tag
```

```lua
tag = "*.0.0" -- Will update last created tag that matches the pattern (e.g: 3.0.0, 4.0.0, 5.0.0,...)
```

**Comment by mvllow** [comment]
Lovely feature! I was testing `tag = 'v1.0.0-canary.*'` but that didn't seem to work. Using `tag = '*'` did use the latest release, although it was a prerelease instead of the latest stable release. Not a huge deal but wanted to share my experience 😌

**Comment by danymat** [comment]
@mvllow what is the output of `git tag -l v1.0.0-canary.*` ?

**Comment by mvllow** [comment]
@danymat
```
No matches for wildcard 'v1.0.0-canary.*'. See `help expand`.
git tag -l v1.0.0-canary.*
```

Whereas `git tag -l v0.5.0` produces no output (another valid branch – referencing rose-pine/neovim repo)

Oddly, `git tag -l v0.*.0` also says no matches so could be user error

**Comment by wbthomason** [comment]
Thanks @danymat for implementing #809!

**Comment by gegoune** [comment]
There also is release-please Github Action which might have better support, not sure if replicating efforts make sense, but good job anyways!

**Comment by danymat** [comment]
The difference in my take is:
- We're not creating release, but just tags
- We don't use commit names for release, but code in your lua plugin

But yeah, those two ideas will work with tag wildcard
